### PR TITLE
Update high level multiplayer manual for Godot 4.

### PR DESCRIPTION
**Pull request by Jordyfel**
Updates the high level multiplayer manual for Godot 4.

I have used the class reference, blog posts and my experience as a user of the API to get the information. I expect to have missed some details as I don't know the internals of `MultiplayerAPI`.

### CODE REVIEW

File: tutorials/networking/high_level_multiplayer.rst
Comment (AThousandShips) [inline tutorials/networking/high_level_multiplayer.rst:106]: Misleading, unless defined for the sub-tree it defaults to the default one, but this implies it's always a default one

File: tutorials/networking/high_level_multiplayer.rst
Comment (AThousandShips) [inline tutorials/networking/high_level_multiplayer.rst:117]: Again, only if none has been configured, this is going to confuse people

File: tutorials/networking/high_level_multiplayer.rst
Comment (AThousandShips) [inline tutorials/networking/high_level_multiplayer.rst:109]:
```suggestion
which will override ``multiplayer`` for the node at that path and all of it's descendants.
```
Not limited to just children

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst:109]: What is the difference? I don't believe I've come across this term in Godot before.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst:109]: "Children" would imply just the direct children, and not grandchildren, great-grandchildren, etc.

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst:109]: Makes sense.

File: tutorials/networking/high_level_multiplayer.rst
Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst:117]: That's what I put the "by default" for, but I can see it not being clear enough. I'm open to suggestions for better language since I'm not too sure.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst:117]: Why is this relevant? It's only true for a very specific case, so it's unhelpful information

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst:117]: I wanted to specifically mention that `Node.multiplayer` is a reference to a "global" object, and not one unique per node. I've seen people ask why you cannot set this property even though you can get it. When learning the API I was myself confused about why one can just get multiplayer from every node and receive the same object.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst:117]: Then the correct code is `get_tree().get_multiplayer(get_path())`

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst:117]: I believe this should stay mentioned, otherwise the example I placed below is confusing. Why is `multiplayer` getting set in a singleton, but works for every node? I understand it's documented in the class reference, but I think it's worth for new users to see.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst:117]: Sure, but the current wording is far too confusing and will cause people to assume this is the correct way of getting the `MultiplayerAPI` of a `Node`, which is bad

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst:117]: I did not understand which is the correct way. I assume it's through the `multiplayer` property, because why would it exist otherwise.

It makes sense to me that the preferred way should be mentioned in the docs.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst:117]: The correct way is `multiplayer`, it ensures you are in the tree

However calling `get_tree().get_multiplayer()` returns the default one, not the one associated with any one node

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst]: Edited these to be more clear.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst]: This is still going to confuse people, why not use the proper equal expression as an example?

Comment (Jordyfel) [reply tutorials/networking/high_level_multiplayer.rst]: I'm not willing to die on this hill... for now I want to see if a networking maintainer approves of the content generally before I put more effort into polishing it.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst]: I'm sorry if this was unpleasant, I didn't mean for it to become an argument, my apologies

**Comment by mhilbrunner** [comment]
@Jordyfel Thanks for working on this, I've work in progress on expanding and reworking the networking sections, but as that turned out to take longer than anticipated I had planned to do a quicker update of the existing ones - so this is very welcome.

I'm currently building this page locally and will be reviewing it, thanks for contributing to Godot, especially to networking docs ❤️

**Review by mhilbrunner — APPROVED** [review]
Great work! I've finally found time to review this in depth. This is a great update to bring this page up to date for Godot 4.

I've found a few things, but these may as well be worth doing in followup PRs (by you, me or someone else):

1. The warning about the Android INTERNET permission was removed in this PR. Was that intentional?
2. Where it explains the signals, the signal names should probably be links to their class docs.
3. I think the flow of the page could be improved (that has nothing to do with this PR though, it was already the case before): It starts with a lot of very high level general networking talk, then gets nitty gritty, and kinda goes back and forth and thats hard to follow. We should probably split the general networking overview, security considerations, comparing Godot's networking technologies etc. from the actual "this is how you use MultiplayerAPI".

It also needs advanced tutorials for the synchronizer tech etc, but that is out of scope for this PR.

All in all, I'll likely merge this sooner rather than later and iterate based on it, as I have further changes in my own fork and will work on the above (further PRs are of course welcome!)

Because as of right now, the current state of the page is badly outdated, so I'm in favor of improving it and then improving it further, rather than spending more time before merging. :)

**Comment by Jordyfel** [comment]
I agree with everything you mentioned, this page probably has to be split/reorganized, though I'm not very confident in doing that since my networking knowledge is rather surface level. I just wanted to have all the basics in one place sooner rather than later because this API, more than most other, is a pain to figure out from the class reference as a new user (not so much the case for the scene replication nodes, which is why I didn't include them, though it would be better if they had tutorials as well).

I didn't notice the android permission warning, will put it back as it was.
I noticed a lot of class/method names could use links to the class reference, I just hadn't figured out how to do it when I was writing. Will try to get it done soon, or if you decide to merge before that, in a separate PR.

File: tutorials/networking/high_level_multiplayer.rst
Comment (FrederickKDP) [reply tutorials/networking/high_level_multiplayer.rst]: I find  `get_tree().get_multiplayer()` to be overly verbose. `get_multiplayer()` works on its own and makes more sense in my opinion. This is important for a tutorial.

Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst]: Those are not equivalent, as is the entire point of this discussion here

Comment (FrederickKDP) [reply tutorials/networking/high_level_multiplayer.rst]: > Those are not equivalent, as is the entire point of this discussion here

I'm not saying that they are. I'm saying I'm not fond of using `get_tree().get_multiplayer()` for the reasons explained. If people really believe it's a must, I believe further explanation is necessary. Because otherwise, people will have the same doubts that we are having in this discussion.
Reasons to use `get_tree().get_multiplayer()` should be clear if used in my opinion.

**Comment by FrederickKDP** [comment]
> I agree with everything you mentioned, this page probably has to be split/reorganized, though I'm not very confident in doing that since my networking knowledge is rather surface level. I just wanted to have all the basics in one place sooner rather than later because this API, more than most other, is a pain to figure out from the class reference as a new user (not so much the case for the scene replication nodes, which is why I didn't include them, though it would be better if they had tutorials as well).
> 
> I didn't notice the android permission warning, will put it back as it was. I noticed a lot of class/method names could use links to the class reference, I just hadn't figured out how to do it when I was writing. Will try to get it done soon, or if you decide to merge before that, in a separate PR.

Agreed. I was myself in this situation. It's very painful to figure it out as a novice.
I would argue that a point to have here is how in-depth should this page be. Is it for novices, or for pros. If is for newcomers, how pros will find clarification? If it tries to be for both, how to explain that without overly complicate for newbies?

File: tutorials/networking/high_level_multiplayer.rst
Comment (AThousandShips) [reply tutorials/networking/high_level_multiplayer.rst]: I'm sorry your original comment made it seem that way based on the suggestion of replacing one with the other

Comment (FrederickKDP) [reply tutorials/networking/high_level_multiplayer.rst]: I see, no worries. It was confusing, the way I put it. My fault as well.
Always please feel free to argue my posts. I'm new here, so I appreciate a lot the feedback. Hopefully I may contribute to the discussion.

**Comment by mhilbrunner** [comment]
Merged! Thanks and congrats @Jordyfel, this is amazing, doubly so for a first contribution :)

> I would argue that a point to have here is how in-depth should this page be

Absolutely. We _strive_ to keep things as welcoming as reasonably possible, but we absolutely do not try to teach people game development (or network engineering), that is way out of scope of the docs and our resources. The Godot's are meant to be a technical manual for this specific game engine; creating in-depth networking/multiplayer tutorials is work for the greater community, we can only manage to document the tech.

**Comment by Jordyfel** [comment]
Other issues I found which can seemingly also be closed by this: #5754, #7142. @mhilbrunner

Some guy on reddit complained that the issue he was tracking wasn't closed.

**Comment by mhilbrunner** [comment]
Thanks for the ping, I closed those issues now :)

**Pull request merged**
